### PR TITLE
Merge/foundation release/1.13.14+15

### PR DESCRIPTION
**Pull request by meowsbits**
Supersedes #621 

---

## What's Changed

- bump min Go version to 1.21

## Fixed

- 

---

Comparison with last release: [v1.13.1..v1.12.19](https://github.com/etclabscore/core-geth/compare/v1.13.1..v1.12.19)
Docker images published under [etclabscore/core-geth](https://hub.docker.com/r/etclabscore/core-geth/builds).

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+  workflow_dispatch:
+
+jobs:
+  build:
+    runs-on: self-hosted
```
Comment (diega) [inline .github/workflows/go.yml:12]: we would probably need to configure this runner on our side

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,168 @@
+language: go
+go_import_path: github.com/ethereum/go-ethereum
```
Comment (diega) [inline .travis.yml:2]: besides changing this, we could try to configure TravisCI to see if we can match some of this builds

File: Dockerfile.alltools [hunk 3]
```diff
@@ -4,7 +4,7 @@ ARG VERSION=""
 ARG BUILDNUM=""
 
 # Build Geth in a stock Go builder container
-FROM golang:1.20-alpine as builder
+FROM golang:1.21-alpine as builder
```
Comment (diega) [inline Dockerfile.alltools]: Add to notes: bump min Go version to 1.21

File: cmd/devp2p/README.md [hunk 4]
```diff
@@ -108,31 +108,32 @@ Start the test by running `devp2p discv5 test -listen1 127.0.0.1 -listen2 127.0.
...
+
+    geth \
+        --datadir <datadir>            \
+        --nodiscover                   \
+        --nat=none                     \
+        --networkid 3503995874084926   \
+        --verbosity 5                  \
+        --authrpc.jwtsecret 0x7365637265747365637265747365637265747365637265747365637265747365
+
+Note that the tests also require access to the engine API.
```
Comment (diega) [inline cmd/devp2p/README.md:126]: Based on this note I'm not sure if we should integrate `devp2p` changes without checking if it still works for our diagnosing usage

File: cmd/devp2p/rlpxcmd.go [hunk 5]
```diff
@@ -46,22 +47,30 @@ var (
 	}
 	rlpxEthTestCommand = &cli.Command{
 		Name:      "eth-test",
-		Usage:     "Runs tests against a node",
-		ArgsUsage: "<node> <chain.rlp> <genesis.json>",
+		Usage:     "Runs eth protocol tests against a node",
```
Comment (diega) [inline cmd/devp2p/rlpxcmd.go:50]: AFAIK we're not using these, but worth to highlight that we'll not be able to use them ever now

File: cmd/evm/runner.go [hunk 6]
```diff
@@ -146,11 +146,11 @@ func runCmd(ctx *cli.Context) error {
 			initialGas = genesisConfig.GasLimit
 		}
 	} else {
-		genesisConfig.Config = params.AllEthashProtocolChanges
+		genesisConfig.Config = params.AllDevChainProtocolChanges
```
Comment (diega) [inline cmd/evm/runner.go:149]: would they be wiping out every _ethash_ concept by replacing them with _dev_?

File: cmd/geth/main.go
Comment (diega) [inline cmd/geth/main.go:1]: LES finally deprecated

File: cmd/utils/flags.go [hunk 7]
```diff
@@ -1543,58 +1545,24 @@ func SetP2PConfig(ctx *cli.Context, cfg *p2p.Config) {
...
-
-	lightPeers := ctx.Int(LightMaxPeersFlag.Name)
-	if lightClient && !ctx.IsSet(LightMaxPeersFlag.Name) {
-		// dynamic default - for clients we use 1/10th of the default for servers
-		lightPeers /= 10
-	}
-
 	if ctx.IsSet(MaxPeersFlag.Name) {
 		cfg.MaxPeers = ctx.Int(MaxPeersFlag.Name)
-		if lightServer && !ctx.IsSet(LightMaxPeersFlag.Name) {
```
Comment (ziogaschr) [inline cmd/utils/flags.go:1557]: Once we release this version, we have to update our nodes to stop serving light clients.

File: cmd/utils/flags.go [hunk 8]
```diff
@@ -2272,11 +2245,26 @@ func SetEthConfig(ctx *cli.Context, stack *node.Node, cfg *ethconfig.Config) {
...
 
 		// Create a new developer genesis block or reuse existing one
-		cfg.Genesis = params.DeveloperGenesisBlock(ctx.Uint64(DeveloperGasLimitFlag.Name), developer.Address, ctx.Bool(DeveloperPoWFlag.Name))
+		cfg.Genesis = params.DeveloperGenesisBlock(ctx.Uint64(DeveloperGasLimitFlag.Name), &developer.Address, ctx.Bool(DeveloperPoWFlag.Name))
 		if ctx.IsSet(DataDirFlag.Name) {
 			chaindb := tryMakeReadOnlyDatabase(ctx, stack)
 			if rawdb.ReadCanonicalHash(chaindb, 0) != (common.Hash{}) {
 				cfg.Genesis = nil // fallback to db content
+
+				// validate genesis has PoS enabled in block 0
```
Comment (ziogaschr) [inline cmd/utils/flags.go:2254]: We need to test `--dev` mode?

File: consensus/ethash/ethash_test.go [hunk 9]
```diff
@@ -27,16 +27,15 @@ import (
...
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/core/types"
-	"github.com/ethereum/go-ethereum/log"
 )
 
 func verboseLogging() {
-	glogger := log.NewGlogHandler(log.StreamHandler(os.Stdout, log.TerminalFormat(false)))
-	glogger.Verbosity(log.Lvl(99))
-	log.Root().SetHandler(glogger)
+	slog.SetDefault(slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelDebug})))
```
Comment (ziogaschr) [inline consensus/ethash/ethash_test.go:38]: Why we don't use the `go-ethereum/log` here?

File: core/genesis.go [hunk 10]
```diff
@@ -98,15 +100,9 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *trie.Database, gen
...
-			if overrides != nil && overrides.OverrideShanghai != nil {
-				config.SetEIP3651TransitionTime(overrides.OverrideShanghai)
-				config.SetEIP3855TransitionTime(overrides.OverrideShanghai)
-				config.SetEIP3860TransitionTime(overrides.OverrideShanghai)
-				config.SetEIP4895TransitionTime(overrides.OverrideShanghai)
-				config.SetEIP6049TransitionTime(overrides.OverrideShanghai)
-			}
 			if overrides != nil && overrides.OverrideCancun != nil {
 				config.SetEIP4844TransitionTime(overrides.OverrideCancun)
+				// TODO(meowsbits) Install the remaining Cancun EIP overrides.
```
Comment (ziogaschr) [inline core/genesis.go]: Note to not forget it

File: consensus/ethash/ethash_test.go (hunk 9 above)
Comment (ziogaschr) [reply consensus/ethash/ethash_test.go:38]: "golang.org/x/exp/slog" is used in the log package either way. But can't we use that instead and keep the dependency just there?

File: params/types/ctypes/configurator_iface.go [hunk 11]
```diff
@@ -148,58 +149,73 @@ type ProtocolSpecifier interface {
 	GetEIP2315Transition() *uint64
 	SetEIP2315Transition(n *uint64) error
 
-	// ModExp gas cost
+	// Berlin:
+
+	// GetEIP2565Transition implements EIP-2565: ModExp Gas Cost - https://eips.ethereum.org/EIPS/eip-2565
```
Comment (ziogaschr) [inline params/types/ctypes/configurator_iface.go:154]: @meowsbits love you for the comments :)

File: cmd/faucet/faucet.go [hunk 12]
```diff
@@ -1,1244 +0,0 @@
-// Copyright 2017 The go-ethereum Authors
```
Comment (ziogaschr) [inline cmd/faucet/faucet.go:1]: @meowsbits do we want to keep and maintain faucet as you were using it?

File: eth/downloader/downloader_test.go [hunk 13]
```diff
@@ -752,9 +723,6 @@ func testMultiSynchronisation(t *testing.T, protocol uint, mode SyncMode) {
 func TestMultiProtoSynchronisation68Full(t *testing.T)  { testMultiProtoSync(t, eth.ETH68, FullSync) }
 func TestMultiProtoSynchronisation68Snap(t *testing.T)  { testMultiProtoSync(t, eth.ETH68, SnapSync) }
 func TestMultiProtoSynchronisation68Light(t *testing.T) { testMultiProtoSync(t, eth.ETH68, LightSync) }
-func TestMultiProtoSynchronisation67Full(t *testing.T)  { testMultiProtoSync(t, eth.ETH67, FullSync) }
```
Comment (ziogaschr) [inline eth/downloader/downloader_test.go:755]: Do we want to keep ETH67 and its tests?

Comment (ziogaschr) [reply eth/downloader/downloader_test.go:755]: What is the latest version implemented by besu and Erigon?

Comment (ziogaschr) [reply eth/downloader/downloader_test.go:755]: This is the PR removing it https://github.com/ethereum/go-ethereum/pull/28956

File: Dockerfile.alltools (hunk 3 above)
Comment (ziogaschr) [reply Dockerfile.alltools]: `go.mod` still uses go1.20. Did this mismatch came from upstream?

**Comment by ziogaschr** [comment]
I have finished with the review and made some comments. After we check/work on all the comments here, we can run a sync test and merge it I think.

File: Dockerfile.alltools (hunk 3 above)
Comment (ziogaschr) [reply Dockerfile.alltools]: Fixed by [b9eb08a](https://github.com/etclabscore/core-geth/pull/623/commits/b9eb08a87b113443110930fc3537832e7a6631c3) and [86f3184](https://github.com/etclabscore/core-geth/pull/623/commits/86f3184415f73e5db5b9e30767d900ce00ba8a13).

Comment (ziogaschr) [reply Dockerfile.alltools]: > Add to notes: bump min Go version to 1.21

I added it in this PR description so as to keep it for later

File: cmd/devp2p/rlpxcmd.go (hunk 5 above)
Comment (ziogaschr) [reply cmd/devp2p/rlpxcmd.go:50]: Nice catch, as I used it once in past.
This haven't dropped support for them, just changed the "ArgsUsage".

File: cmd/evm/runner.go (hunk 6 above)
Comment (ziogaschr) [reply cmd/evm/runner.go:149]: They will actually continue supporting only post-merge chains

File: cmd/utils/flags.go (hunk 8 above)
Comment (ziogaschr) [reply cmd/utils/flags.go:2254]: `--dev.pow` works. It mines blocks really fast, do you remember if it was waiting some time between blocks by default?

File: core/genesis.go (hunk 10 above)
Comment (ziogaschr) [reply core/genesis.go]: Fixed with [9290ad9](https://github.com/etclabscore/core-geth/pull/623/commits/9290ad92a48c0a810bfa8916da8646e4cd77a462)

File: eth/protocols/eth/handler.go
Comment (ziogaschr) [inline eth/protocols/eth/handler.go:1]: We can probably remove this comment https://github.com/etclabscore/core-geth/pull/623/files#diff-5beb4052baf94e3025566d929c63974350e4b1eae18cce8eb21b4579d6788ffeR98.
@meowsbits is there anything else to do about it?

File: eth/downloader/downloader_test.go (hunk 13 above)
Comment (ziogaschr) [reply eth/downloader/downloader_test.go:755]: PR that reverts the ETH67 removal https://github.com/etclabscore/core-geth/pull/629

File: cmd/utils/flags.go (hunk 8 above)
Comment (meowsbits) [reply cmd/utils/flags.go:2254]: It's too fast, like 1000 blocks / 10 seconds. Its probably using `--fakepow` with no pause. I'll take a look at it.

File: cmd/faucet/faucet.go (hunk 12 above)
Comment (meowsbits) [reply cmd/faucet/faucet.go:1]: I was using it for Mordor. If we have, or there are, other ways of distributing tokens then I'm fine abandoning or archiving it.

File: cmd/utils/flags.go (hunk 8 above)
Comment (meowsbits) [reply cmd/utils/flags.go:2254]: Actually... I only get the super-fast blocks when I use `./build/bin/geth --dev.pow --fakepow.poisson`. If I just run it simply with `./build/bin/geth --dev.pow`, after startup geth waits for me to `geth attach /tmp/geth.ipc` and run `miner.start()`, and then it produces blocks at a reasonable pace. So I think if anything's broken here, it's probably the `--fakepow.poisson` which wasn't exactly designed to be used with developer mode anyways.

I think there's nothing to fix.

File: eth/protocols/eth/handler.go
Comment (meowsbits) [reply eth/protocols/eth/handler.go:1]: My browser is struggling to load the linked comment. Please clarify which comment you're pointing to.

File: cmd/faucet/faucet.go (hunk 12 above)
Comment (ziogaschr) [reply cmd/faucet/faucet.go:1]: @diega I think you also have a faucet. Which backend does it use?

File: eth/downloader/downloader_test.go (hunk 13 above)
Comment (ziogaschr) [reply eth/downloader/downloader_test.go:755]: Ohh, @meowsbits @diega we have to merge this one I think. If yes, then we have to rerun our sync test.

File: eth/protocols/eth/handler.go
Comment (ziogaschr) [reply eth/protocols/eth/handler.go:1]: My browser can't load it either. It was probably a comment in code with a TODO on your name.

File: cmd/faucet/faucet.go (hunk 12 above)
Comment (diega) [reply cmd/faucet/faucet.go:1]: It's just a CF worker that gives away 1 mETC to any address with less than 2 mETC that I fill from the mining rewards of the Besu node that I maintain

**Pull request merged**